readme.md
<img src="https://wedgetail.netlify.com/logo.png" height=100 width=100 />

# Wedgetail

_Performance test your functions_

[![View on npm](https://img.shields.io/npm/v/wedgetail.svg)](https://npmjs.org/packages/wedgetail)
[![npm downloads](https://img.shields.io/npm/dm/wedgetail.svg)](https://npmjs.org/packages/wedgetail)
[![Dependencies](https://img.shields.io/david/ojkelly/wedgetail.svg)](https://david-dm.org/ojkelly/wedgetail)
[![Build Status](https://travis-ci.org/ojkelly/wedgetail.svg?branch=master)](https://travis-ci.org/ojkelly/wedgetail)
[![codecov](https://codecov.io/gh/ojkelly/wedgetail/branch/master/graph/badge.svg)](https://codecov.io/gh/ojkelly/wedgetail)
[![NSP Status](https://nodesecurity.io/orgs/ojkelly/projects/7f441bdb-76ab-4155-aec9-00777b5adc9a/badge)](https://nodesecurity.io/orgs/ojkelly/projects/7f441bdb-76ab-4155-aec9-00777b5adc9a)[![Known Vulnerabilities](https://snyk.io/test/npm/wedgetail/badge.svg)](https://snyk.io/test/npm/wedgetail)

Wedgetail is a small performance tesing library that runs on NodeJS `9.5.0` and above. It's designed
to be used inside your tests, to ensure your desired function is always as fast as you want it to be.

## Getting Started

Add `wedgetail` to your `devDependencies`.

`yarn add -D wedgetail`

`npm install --save-dev wedgetail`

`Wedgetail` can be easily inserted into your tests. It's reccomended you create a seperate test
just for performance, as you cannot use the result of the function you are testing.

The function you are testing can also be your whole test.

### Usage

To use `wedgetail` you need to call in inside the callback.

```javascript
// javascript
import test from "ava";

import { timeExecution } from "wedgetail";

test("Can time a function", async t => {
    // This object contains a definition of the threshold
    // at which your function is too slow.
    // All timings are in milliseconds(ms)
    const allowedPerformance = {
        average: 0.001,
        high: 1,
        low: 0.001,
        percentiles: {
            ninetieth: 0.0004,
            ninetyFifth: 0.001,
            ninetyNinth: 0.001,
            tenth: 0.0005,
        },
    };

    const timings = await timeExecution({
        expectedTimings: allowedPerformance,
        numberOfExecutions: 5000,
        // By using an anonymous arrow function you should
        // be able to maintain the correct scope
        // of `this`.
        callback: () => {
            // Your function goes here
            Math.sqrt(Math.random());
        },
    });
    // You can use any testing or assertion library.
    // if the timings are below your expected values then
    // `timings.results.passed` will be `true`
    t.true(timings.results.passed, "timings failed");
});
```

```typescript
// typescript
import test from "ava";

import { timeExecution, Timings, TimedPerformance } from "wedgetail";

test("Can time a function", async t => {
    // This object contains a definition of the threshold
    // at which your function is too slow.
    // All timings are in milliseconds(ms)
    const allowedPerformance: Timings = {
        average: 0.001,
        high: 1,
        low: 0.001,
        percentiles: {
            ninetieth: 0.0004,
            ninetyFifth: 0.001,
            ninetyNinth: 0.001,
            tenth: 0.0005,
        },
    };

    const timings: TimedPerformance = await timeExecution({
        expectedTimings: allowedPerformance,
        numberOfExecutions: 5000,
        // By using an anonymous arrow function you should
        // be able to maintain the correct scope
        // of `this`.
        callback: () => {
            // Your function goes here
            Math.sqrt(Math.random());
        },
    });
    // You can use any testing or assertion library.
    // if the timings are below your expected values then
    // `timings.results.passed` will be `true`
    t.true(timings.results.passed, "timings failed");
});
```

You will need to play around with `numberOfExecutions` to find the right number. If your function is
slow (2ms+) you may want to use a number less than 5000.

## Running the tests

Use `yarn tests` or `npm run tests`.

Tests are written with `ava`, and we would strongly like tests with any new functionality.

## Contributing

Please read [CONTRIBUTING.md](https://github.com/ojkelly/wedgetail/CONTRIBUTING.md) for details on our code of conduct, and the process for submitting pull requests to us.

## Versioning

We use [SemVer](http://semver.org/) for versioning. For the versions available, see the [tags on this repository](https://github.com/ojkelly/wedgetail/tags).

## Authors

* **Owen Kelly** - [ojkelly](https://github.com/ojkelly)

## License

This project is licensed under the MIT License - see the [LICENSE.md](https://github.com/ojkelly/wedgetail/LICENSE.md) file for details


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>
# [1.0.0](https://github.com/ojkelly/wedgetail/compare/v0.9.9...v1.0.0) (2018-02-28)



<a name="0.9.9"></a>
## [0.9.9](https://github.com/ojkelly/wedgetail/compare/v0.9.8...v0.9.9) (2018-02-20)



<a name="0.9.8"></a>
## [0.9.8](https://github.com/ojkelly/wedgetail/compare/v0.9.7...v0.9.8) (2018-02-20)



<a name="0.9.7"></a>
## [0.9.7](https://github.com/ojkelly/wedgetail/compare/v0.9.6...v0.9.7) (2018-02-19)



<a name="0.9.6"></a>
## [0.9.6](https://github.com/ojkelly/wedgetail/compare/v0.9.5...v0.9.6) (2018-02-19)



<a name="0.9.5"></a>
## [0.9.5](https://github.com/ojkelly/wedgetail/compare/v0.9.4...v0.9.5) (2018-02-19)



<a name="0.9.4"></a>
## [0.9.4](https://github.com/ojkelly/wedgetail/compare/v0.9.3...v0.9.4) (2018-02-19)



<a name="0.9.3"></a>
## [0.9.3](https://github.com/ojkelly/wedgetail/compare/v0.9.2...v0.9.3) (2018-02-19)



<a name="0.9.2"></a>
## [0.9.2](https://github.com/ojkelly/wedgetail/compare/v0.9.1...v0.9.2) (2018-02-19)



<a name="0.9.1"></a>
## 0.9.1 (2018-02-19)


tests/index.spec.ts
import test from "ava";

import {
    timeExecution,
    TimeExecutionOptions,
    Timings,
    TimedPerformance,
} from "../src/index";

test("Can time a function", async t => {
    const allowedPerformance: Timings = {
        average: 0.5,
        high: 1,
        low: 0.01,
        percentiles: {
            ninetieth: 0.5,
            ninetyFifth: 0.5,
            ninetyNinth: 0.5,
            tenth: 0.5,
        },
    };
    const timings: TimedPerformance = await timeExecution({
        expectedTimings: allowedPerformance,
        numberOfExecutions: 5000,
        callback: () => {
            Math.sqrt(Math.random());
        },
    });
    t.true(timings.results.passed, "timings failed");
});


lib/tests/index.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ava_1 = require("ava");
const index_1 = require("../src/index");
ava_1.default("Can time a function", (t) => __awaiter(this, void 0, void 0, function* () {
    const allowedPerformance = {
        average: 0.5,
        high: 1,
        low: 0.01,
        percentiles: {
            ninetieth: 0.5,
            ninetyFifth: 0.5,
            ninetyNinth: 0.5,
            tenth: 0.5,
        },
    };
    const timings = yield index_1.timeExecution({
        expectedTimings: allowedPerformance,
        numberOfExecutions: 5000,
        callback: () => {
            Math.sqrt(Math.random());
        },
    });
    t.true(timings.results.passed, "timings failed");
}));
//# sourceMappingURL=index.spec.js.map

gulpfile.js
"use strict";

var gulp = require("gulp");
var gulpIf = require("gulp-if");
var gulpRev = require("gulp-rev");
var gulpRevCollector = require("gulp-rev-collector");
var gulpRevReplace = require("gulp-rev-replace");
var gulpUglify = require("gulp-uglify");
var gulpUniqueFiles = require("gulp-unique-files");
var gulpUseRef = require("gulp-useref");
var gulpCleanCSS = require("gulp-clean-css");
var gulpResponsive = require("gulp-responsive");
var gulpCheerio = require("gulp-cheerio");
var del = require("del");
var rename = require("rename");

var dirs = {
  public: "public",
  screenshots: "public/build/screenshots"
};

gulp.task("useref", function() {
  var assets = gulpUseRef.assets({
    searchPath: "public"
  });

  return gulp
    .src("public/**/*.html")
    .pipe(assets)
    .pipe(gulpUniqueFiles())
    .pipe(gulpIf("*.css", gulpCleanCSS()))
    .pipe(gulpIf("*.js", gulpUglify()))
    .pipe(gulpRev())
    .pipe(assets.restore())
    .pipe(gulpUseRef())
    .pipe(
      gulpRevReplace({
        prefix: "/"
      })
    )
    .pipe(gulp.dest("public"));
});

// gulp.task('screenshot:clean', function() {
//   return del([dirs.screenshots + '/**/*']);
// });

// gulp.task('screenshot:rev', ['screenshot:clean'], function() {
//   return gulp.src('public/themes/screenshots/*.png')
//     .pipe(gulpRev())
//     .pipe(gulp.dest(dirs.screenshots))
//     .pipe(gulpRev.manifest())
//     .pipe(gulp.dest(dirs.screenshots));
// });

// gulp.task('screenshot:revreplace', ['screenshot:rev'], function() {
//   var destDir = '/build/screenshots';

//   return gulp.src([dirs.screenshots + '/rev-manifest.json', 'public/themes/index.html'])
//     .pipe(gulpRevCollector({
//       replaceReved: true,
//       dirReplacements: {
//         '/themes/screenshots': destDir
//       }
//     }))
//     .pipe(gulpCheerio(function($, file) {
//       $('img.plugin-screenshot-img.lazyload').each(function() {
//         var img = $(this);
//         var src = img.attr('data-src') || img.attr('data-org');
//         if (!src) return;

//         var jpgPath = replaceBackSlash(rename(src, {extname: '.jpg'}));
//         var jpg2xPath = replaceBackSlash(rename(jpgPath, {suffix: '@2x'}));
//         var srcset = [
//           jpgPath,
//           jpg2xPath + ' 2x'
//         ].join(', ');

//         img.attr('data-src', jpgPath)
//           .attr('data-srcset', srcset)
//           .attr('data-org', src);
//       });
//     }))
//     .pipe(gulp.dest('public/themes'));
// });

// gulp.task('screenshot:resize', ['screenshot:rev'], function() {
//   return gulp.src(dirs.screenshots + '/*.png')
//     .pipe(gulpResponsive({
//       '*.png': [
//         {
//           width: '50%',
//           rename: {
//             extname: '.jpg'
//           }
//         },
//         {
//           rename: {
//             suffix: '@2x',
//             extname: '.jpg'
//           }
//         }
//       ]
//     }, {
//       progressive: true,
//       format: 'jpeg',
//       quality: 70,
//       stats: false
//     }))
//     .pipe(gulp.dest(dirs.screenshots));
// });

// gulp.task('screenshot', ['screenshot:rev', 'screenshot:resize', 'screenshot:revreplace']);
gulp.task("default", ["useref"]);

function replaceBackSlash(str) {
  return str.replace(/\\/g, "/");
}


src/index.ts
import { performance } from "perf_hooks";

import * as percentile from "percentile";

type TimeExecutionCallback = (...callbackArgs) => any;
type TimeExecutionOptions = {
    expectedTimings: Timings;
    numberOfExecutions: number;
    callback: TimeExecutionCallback;
    callbackArgs?: any[];
};
type Timings = {
    high: number;
    low: number;
    average: number;
    percentiles: {
        ninetyNinth: number;
        ninetyFifth: number;
        ninetieth: number;
        tenth: number;
    };
};
type TimedPerformance = {
    timings: Timings;
    results: {
        passed: boolean;
        high: boolean;
        low: boolean;
        average: boolean;
        percentiles: {
            ninetyNinth: boolean;
            ninetyFifth: boolean;
            ninetieth: boolean;
            tenth: boolean;
        };
    };
};

type ComparePerformanceOutcome = {
    passed: boolean;
    high: boolean;
    low: boolean;
    average: boolean;
    percentiles: {
        ninetyNinth: boolean;
        ninetyFifth: boolean;
        ninetieth: boolean;
        tenth: boolean;
    };
};

async function timeExecution(
    this: any,
    {
        expectedTimings,
        numberOfExecutions,
        callback,
        callbackArgs,
    }: TimeExecutionOptions,
): Promise<TimedPerformance> {
    // Execute the function
    const executions: any[] = await Array(numberOfExecutions)
        .fill("0")
        .map(() => {
            const startTimeMs: number = performance.now();

            callback.apply(this, callbackArgs);

            const endTimeMs: number = performance.now();
            const elapsed: number = endTimeMs - startTimeMs;
            return endTimeMs - startTimeMs;
        });

    const high: number = Math.max.apply(Math, executions);
    const low: number = Math.min.apply(Math, executions);

    // Calc average
    const sum: number = executions.reduce(
        (acc: number, current: number) => acc + current,
    );
    const average: number = sum / numberOfExecutions;
    const timings: any = {
        high,
        low,
        average,
        percentiles: {
            ninetyNinth: percentile(99, executions),
            ninetyFifth: percentile(95, executions),
            ninetieth: percentile(90, executions),
            tenth: percentile(10, executions),
        },
    };
    const results: any = comparePerformance({
        expected: expectedTimings,
        results: timings,
    });

    return {
        timings,
        results,
    };
}

function comparePerformance({
    expected,
    results,
}: {
    expected: Timings;
    results: Timings;
}): ComparePerformanceOutcome {
    let outcome: ComparePerformanceOutcome = {
        passed: false,
        high: expected.high > results.high,
        low: expected.low > results.low,
        average: expected.average > results.average,
        percentiles: {
            ninetyNinth:
                expected.percentiles.ninetyNinth >
                results.percentiles.ninetyNinth,
            ninetyFifth:
                expected.percentiles.ninetyFifth >
                results.percentiles.ninetyFifth,
            ninetieth:
                expected.percentiles.ninetieth > results.percentiles.ninetieth,
            tenth: expected.percentiles.tenth > results.percentiles.tenth,
        },
    };

    if (
        outcome.high &&
        outcome.low &&
        outcome.average &&
        outcome.percentiles.ninetyNinth &&
        outcome.percentiles.ninetyFifth &&
        outcome.percentiles.ninetieth &&
        outcome.percentiles.tenth
    ) {
        outcome.passed = true;
    }
    return outcome;
}

export {
    timeExecution,
    TimeExecutionOptions,
    Timings,
    TimeExecutionCallback,
    TimedPerformance,
    comparePerformance,
    ComparePerformanceOutcome,
};


lib/src/index.d.ts
declare type TimeExecutionCallback = (...callbackArgs) => any;
declare type TimeExecutionOptions = {
    expectedTimings: Timings;
    numberOfExecutions: number;
    callback: TimeExecutionCallback;
    callbackArgs?: any[];
};
declare type Timings = {
    high: number;
    low: number;
    average: number;
    percentiles: {
        ninetyNinth: number;
        ninetyFifth: number;
        ninetieth: number;
        tenth: number;
    };
};
declare type TimedPerformance = {
    timings: Timings;
    results: {
        passed: boolean;
        high: boolean;
        low: boolean;
        average: boolean;
        percentiles: {
            ninetyNinth: boolean;
            ninetyFifth: boolean;
            ninetieth: boolean;
            tenth: boolean;
        };
    };
};
declare type ComparePerformanceOutcome = {
    passed: boolean;
    high: boolean;
    low: boolean;
    average: boolean;
    percentiles: {
        ninetyNinth: boolean;
        ninetyFifth: boolean;
        ninetieth: boolean;
        tenth: boolean;
    };
};
declare function timeExecution(this: any, {expectedTimings, numberOfExecutions, callback, callbackArgs}: TimeExecutionOptions): Promise<TimedPerformance>;
declare function comparePerformance({expected, results}: {
    expected: Timings;
    results: Timings;
}): ComparePerformanceOutcome;
export { timeExecution, TimeExecutionOptions, Timings, TimeExecutionCallback, TimedPerformance, comparePerformance, ComparePerformanceOutcome };


scripts/helpers.js
/* global hexo */

"use strict";

var pathFn = require("path");
var _ = require("lodash");
var cheerio = require("cheerio");
var lunr = require("lunr");

var localizedPath = ["docs", "api"];

function startsWith(str, start) {
  return str.substring(0, start.length) === start;
}

hexo.extend.helper.register("page_nav", function() {
  var type = this.page.canonical_path.split("/")[0];
  var sidebar = this.site.data.sidebar[type];
  var path = pathFn.basename(this.path);
  var list = {};
  var prefix = "sidebar." + type + ".";

  for (var i in sidebar) {
    for (var j in sidebar[i]) {
      list[sidebar[i][j]] = j;
    }
  }

  var keys = Object.keys(list);
  var index = keys.indexOf(path);
  var result = "";

  if (index > 0) {
    result +=
      '<a href="' +
      keys[index - 1] +
      '" class="article-footer-prev" title="' +
      this.__(prefix + list[keys[index - 1]]) +
      '">' +
      '<i class="fa fa-chevron-left"></i><span>' +
      this.__("page.prev") +
      "</span></a>";
  }

  if (index < keys.length - 1) {
    result +=
      '<a href="' +
      keys[index + 1] +
      '" class="article-footer-next" title="' +
      this.__(prefix + list[keys[index + 1]]) +
      '">' +
      "<span>" +
      this.__("page.next") +
      '</span><i class="fa fa-chevron-right"></i></a>';
  }

  return result;
});

hexo.extend.helper.register("doc_sidebar", function(className) {
  var type = this.page.canonical_path.split("/")[0];
  var sidebar = this.site.data.sidebar[type];
  var path = pathFn.basename(this.path);
  var result = "";
  var self = this;
  var prefix = "sidebar." + type + ".";

  _.each(sidebar, function(menu, title) {
    result +=
      '<strong class="' +
      className +
      '-title">' +
      self.__(prefix + title) +
      "</strong>";

    _.each(menu, function(link, text) {
      var itemClass = className + "-link";
      if (link === path) itemClass += " current";

      result +=
        '<a href="' +
        link +
        '" class="' +
        itemClass +
        '">' +
        self.__(prefix + text) +
        "</a>";
    });
  });

  return result;
});

hexo.extend.helper.register("header_menu", function(className) {
  var menu = this.site.data.menu;
  var result = "";
  var self = this;
  var lang = this.page.lang;
  var isEnglish = lang === "en";

  _.each(menu, function(path, title) {
    if (!isEnglish && ~localizedPath.indexOf(title)) path = lang + path;

    result +=
      '<a href="' +
      self.url_for(path) +
      '" class="' +
      className +
      '-link">' +
      self.__("menu." + title) +
      "</a>";
  });

  return result;
});

hexo.extend.helper.register("canonical_url", function(lang) {
  var path = this.page.canonical_path;
  if (lang && lang !== "en") path = lang + "/" + path;

  return this.config.url + "/" + path;
});

hexo.extend.helper.register("url_for_lang", function(path) {
  var lang = this.page.lang;
  var url = this.url_for(path);

  if (lang !== "en" && url[0] === "/") url = "/" + lang + url;

  return url;
});

hexo.extend.helper.register("raw_link", function(path) {
  return (
    "https://github.com/ojkelly/wedgetail/edit/documentation/source/" + path
  );
});

hexo.extend.helper.register("page_anchor", function(str) {
  var $ = cheerio.load(str, { decodeEntities: false });
  var headings = $("h1, h2, h3, h4, h5, h6");

  if (!headings.length) return str;

  headings.each(function() {
    var id = $(this).attr("id");

    $(this)
      .addClass("article-heading")
      .append(
        '<a class="article-anchor" href="#' + id + '" aria-hidden="true"></a>'
      );
  });

  return $.html();
});

hexo.extend.helper.register("lunr_index", function(data) {
  var index = lunr(function() {
    this.field("name", { boost: 10 });
    this.field("tags", { boost: 50 });
    this.field("description");
    this.ref("id");

    _.sortBy(data, "name").forEach((item, i) => {
      this.add(_.assign({ id: i }, item));
    });
  });

  return JSON.stringify(index);
});

hexo.extend.helper.register("canonical_path_for_nav", function() {
  var path = this.page.canonical_path;

  if (startsWith(path, "docs/") || startsWith(path, "api/")) {
    return path;
  }
  return "";
});

hexo.extend.helper.register("lang_name", function(lang) {
  var data = this.site.data.languages[lang];
  return data.name || data;
});

hexo.extend.helper.register("disqus_lang", function() {
  var lang = this.page.lang;
  var data = this.site.data.languages[lang];

  return data.disqus_lang || lang;
});

hexo.extend.helper.register("hexo_version", function() {
  return this.env.version;
});


lib/src/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const perf_hooks_1 = require("perf_hooks");
const percentile = require("percentile");
function timeExecution({ expectedTimings, numberOfExecutions, callback, callbackArgs, }) {
    return __awaiter(this, void 0, void 0, function* () {
        const executions = yield Array(numberOfExecutions)
            .fill("0")
            .map(() => {
            const startTimeMs = perf_hooks_1.performance.now();
            callback.apply(this, callbackArgs);
            const endTimeMs = perf_hooks_1.performance.now();
            const elapsed = endTimeMs - startTimeMs;
            return endTimeMs - startTimeMs;
        });
        const high = Math.max.apply(Math, executions);
        const low = Math.min.apply(Math, executions);
        const sum = executions.reduce((acc, current) => acc + current);
        const average = sum / numberOfExecutions;
        const timings = {
            high,
            low,
            average,
            percentiles: {
                ninetyNinth: percentile(99, executions),
                ninetyFifth: percentile(95, executions),
                ninetieth: percentile(90, executions),
                tenth: percentile(10, executions),
            },
        };
        const results = comparePerformance({
            expected: expectedTimings,
            results: timings,
        });
        return {
            timings,
            results,
        };
    });
}
exports.timeExecution = timeExecution;
function comparePerformance({ expected, results, }) {
    let outcome = {
        passed: false,
        high: expected.high > results.high,
        low: expected.low > results.low,
        average: expected.average > results.average,
        percentiles: {
            ninetyNinth: expected.percentiles.ninetyNinth >
                results.percentiles.ninetyNinth,
            ninetyFifth: expected.percentiles.ninetyFifth >
                results.percentiles.ninetyFifth,
            ninetieth: expected.percentiles.ninetieth > results.percentiles.ninetieth,
            tenth: expected.percentiles.tenth > results.percentiles.tenth,
        },
    };
    if (outcome.high &&
        outcome.low &&
        outcome.average &&
        outcome.percentiles.ninetyNinth &&
        outcome.percentiles.ninetyFifth &&
        outcome.percentiles.ninetieth &&
        outcome.percentiles.tenth) {
        outcome.passed = true;
    }
    return outcome;
}
exports.comparePerformance = comparePerformance;
//# sourceMappingURL=index.js.map